services/business/checkers/virtual_services/route_checker.go

```diff
@@ -0,0 +1,97 @@
+package virtual_services
+
+import (
+	"reflect"
+	"strconv"
+
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/services/models"
+	"github.com/kiali/kiali/util/intutil"
+)
+
+type RouteChecker struct{ kubernetes.IstioObject }
+
+// Check returns both an array of IstioCheck and a boolean indicating if the current route rule is valid.
+// The array of IstioChecks contains the result of running the following validations:
+// 1. All weights with a numeric number.
+// 2. All weights have value between 0 and 100.
+// 3. Sum of all weights are 100 (if only one weight, then it assumes that is 100).
+// 4. All the route has to have weight label.
+func (route RouteChecker) Check() ([]*models.IstioCheck, bool) {
+	checks, valid := route.checkRoutesFor("http")
+	tcpChecks, tcpValid := route.checkRoutesFor("tcp")
+	return append(checks, tcpChecks...), valid && tcpValid
+}
+
+func (route RouteChecker) checkRoutesFor(kind string) ([]*models.IstioCheck, bool) {
+	validations := make([]*models.IstioCheck, 0)
+
+	var weightSum int
+	var weightCount int
+	var valid = true
+
+	http := route.GetSpec()[kind]
+	if http == nil {
+		return validations, valid
+	}
+
+	// Getting a []HTTPRoute
+	slice := reflect.ValueOf(http)
```

Comment:
I guess all of these reflections are somehow due to the lack of generics in go? That makes me feel uncomfortable, but maybe there's no other simple way.

Reply:
Exactly. I find myself uncomfortable too. I tried to cast the whole data to its real shape (istio.HTTPRoute) but it didn't work. The feel I have here is that I spend tons of lines of code casting from interface{} to real data. And in every cast, then make the validations. It is (really)^100 annoying.